File: roblox/Starline/InputNoise.lua
--!strict
--[[ No false reward: ∂c/∂noise < 0 ]]
local Config = require(script.Parent.Config)

local InputNoise = {}

local function C(cfg: any?): any
	return cfg or Config
end

function InputNoise.metric(state: any, cfg: any?): number
	local c = C(cfg)
	local dsteer = math.abs(state.steer - (state.steerPrev or 0))
	local oscillation = dsteer * math.abs(state.steer)
	local brakeSpam = state.brake * dsteer
	local jerkNoise = math.min(1, math.abs(state.jerk) / math.max(c.JerkNoiseScale, 1e-6))
	return c.NoiseSteerWeight * dsteer
		+ c.NoiseOscillationWeight * oscillation
		+ c.NoiseBrakeWeight * brakeSpam
		+ c.NoiseJerkWeight * jerkNoise
end

return InputNoise

File: python/src/starline/input_noise.py
"""Input noise metric — enforces ∂c/∂noise < 0 (no false reward)."""

from __future__ import annotations

from .config import EngineConfig
from .player import PlayerState


def input_noise(p: PlayerState, cfg: EngineConfig) -> float:
    """Dimensionless noise ∈ [0, ~1+] from incoherent control.

    Captures: rapid steer flips, brake spam, oscillation exploits.
    """
    dsteer = abs(p.steer - p.steer_prev)
    oscillation = dsteer * abs(p.steer)  # exploit: sign-flip spam
    brake_spam = p.brake * dsteer
    jerk_noise = min(1.0, abs(p.jerk) / max(cfg.jerk_noise_scale, 1e-6))
    return (
        cfg.noise_steer_weight * dsteer
        + cfg.noise_oscillation_weight * oscillation
        + cfg.noise_brake_weight * brake_spam
        + cfg.noise_jerk_weight * jerk_noise
    )
